### PR TITLE
Add time invariant validation functions

### PR DESCRIPTION
**Pull request by jared-paul**
### Description
Added a whole bunch of validation functions to be used in the simulated tests.

### Testing Done

Need to write tests that use these functions to see if they are correct

### Resolved Issues

#1374 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Good work, I'll review in detail later. Can you take a look at integrating [ssl-game-controller](https://github.com/RoboCup-SSL/ssl-game-controller), and in particular if [game events](https://github.com/RoboCup-SSL/ssl-game-controller/blob/master/proto/ssl_gc_game_event.proto) can be used for validation? I think that a lot of these game events will be useful, but they may not provide as much detail or granularity as our own validation functions

**Comment by MathewMacDougall** [comment]
> Good work, I'll review in detail later. Can you take a look at integrating ssl-game-controller, and in particular if game events can be used for validation? I think that a lot of these game events will be useful, but they may not provide as much detail or granularity as our own validation functions

I think gamecontroller integration is quite out of scope for this PR?

**Comment by jonathanlew** [comment]
> I think gamecontroller integration is quite out of scope for this PR?

I think we should figure which validation functions we can implement using `ssl-gamecontroller` and _not_ implement those ones in this PR

**Comment by garethellis0** [comment]
Um, apologies but I'm confused here? Why not just review and merge the validation functions Jared has added and then add more in a future PR?

File: src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.cpp:19]: Nit; EOF missing.

File: src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.h
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.h:7]: EOF missing.

File: src/software/simulated_tests/validation_functions/ball_velocity_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/ball_velocity_validation.cpp:17]: EOF missing.

File: src/software/simulated_tests/validation_functions/distance_to_ball_kick_off_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/distance_to_ball_kick_off_validation.cpp:32]: EOF missing.

File: src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp:33]: EOF missing.

File: src/software/simulated_tests/validation_functions/ball_velocity_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/ball_velocity_validation.cpp:10]: The function is called `ballExceedsVelocity`, but if the `ball_velocity.length()` exceeds 6.5, it returns false. This seems counter-intuitive to the function name. Consider flipping it around.

File: src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp:18]: Where does this `abs_error` value come from? The comment says it robots must not be < 0.5 meters; is this related?

File: src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp:12]: Same question about where 0.05 came from. Commenting would be helpful here.

File: src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp:41]: EOF missing.

**Comment by jonathanlew** [comment]
> Um, apologies but I'm confused here? Why not just review and merge the validation functions Jared has added and then add more in a future PR?

To be clear, Jared plans to add tests to these functions and I think he's asking for an initial round of review now. It'd be less work for him if we can decide that some or all of the validation functions can leverage ssl-gamecontroller so he won't have to write tests for those.

File: src/software/simulated_tests/validation_functions/one_meter_from_ball_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/one_meter_from_ball_validation.cpp:24]: Missing EOF.

File: src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp:55]: Missing EOF.

File: src/software/simulated_tests/validation_functions/robot_speed_during_stoppage_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/robot_speed_during_stoppage_validation.cpp:16]: Likewise here, seems counter-intuitive to the function name.

File: src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.cpp
Comment (kelvinkoon) [inline src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.cpp:37]: Missing EOF.

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
:timer_clock: 
Functions look solid overall, just a few questions/nits.

File: src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.cpp:12]: This will return `true` if the ball is contained in the field lines, meaning the ball is _not_ out of bounds

File: src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/ball_out_of_bounds_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/ball_velocity_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/ball_velocity_validation.cpp:10]: We have a BALL_MAX_SPEED constant that can be used here

File: src/software/simulated_tests/validation_functions/ball_velocity_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/ball_velocity_validation.h:6]: please add a javadoc comment for this function

Also to help be clear we're testing against the max allowed velocity by the rules (not a velocity provided as a parameter), perhaps a more descriptive name would be `ballExceedsMaxVelocity`

File: src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp:26]: I don't think robots would be considered "too close" if it's not the enemy team's free kick?

File: src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp
Comment (MathewMacDougall) [reply src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.cpp:18]: I'm also confused here. Also the rule defines the gap must be 0.5m, I believe from the _edge of the robot_, not the center, so we likely need to take robot and ball radius into account here (but please double check the rules).

File: src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/distance_to_ball_free_kick_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/distance_to_ball_kick_off_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/distance_to_ball_kick_off_validation.cpp:17]: similar comments to above with the free kick

File: src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp
Comment (MathewMacDougall) [reply src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp:12]: Also need to take radii into account here

File: src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp:20]: this isn't using the lambda function defined above?

File: src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp:14]: We actually have a `getAllRobotsExceptGoalie()` function that can be used here

File: src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/one_meter_from_ball_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/one_meter_from_ball_validation.h:6]: I don't recall any rules where we need to be one meter from the ball. What will this be used for?

File: src/software/simulated_tests/validation_functions/robot_colliding_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_colliding_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp:9]: variable name doesn't match what it's doing

File: src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp:25]: IDs may conflict between enemy and friendly teams, so this wont' catch collisions between a friendly and enemy robot with the same ID

Edit: For pairwise iteration it's probably easier and more readable to just do a nested loop with indices or iterators ([like this](https://codereview.stackexchange.com/a/23287))

File: src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp:41]: Just to be clear, this for returning the robot that would get the foul? (ie. the faster one?)

File: src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_colliding_validation.cpp:48]: Why are we returning `true` here? I don't see any cases where this function ever returns `false`?

File: src/software/simulated_tests/validation_functions/robot_speed_during_stoppage_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_speed_during_stoppage_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.h
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.h:6]: please add a javadoc comment for this function

File: src/software/simulated_tests/validation_functions/robot_speed_during_stoppage_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/robot_speed_during_stoppage_validation.cpp:15]: TODO should be resolved or linked to a ticket before this merges

File: src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.cpp:6]: may want to specify that no _friendly_ robot touches the ball in the enemy defense

File: src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/validation_functions/touch_ball_in_enemy_defense_validation.cpp:25]: Seems like this should return true because the robot _is_ touching the ball here?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Glad to see these coming together!

File: src/software/simulated_tests/validation_functions/distance_to_ball_kick_off_validation.cpp
Comment (EvanMorcom) [inline src/software/simulated_tests/validation_functions/distance_to_ball_kick_off_validation.cpp:25]: Pretty sure this returns _true_ if the the enemy is not taking a kick-off

File: src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp
Comment (EvanMorcom) [inline src/software/simulated_tests/validation_functions/enemy_defense_ball_touch_validation.cpp:3]: This is check if a friendly robot has touched the ball _in the enemy defense_, but there is no mention of the defensive area.

It is possible that this may mistake any touches as being in the defensive area

File: src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp
Comment (EvanMorcom) [inline src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp:12]: There are a couple strange '_aabb' variable names here.

I assume that's a bad refactor?

File: src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp
Comment (EvanMorcom) [inline src/software/simulated_tests/validation_functions/non_keeper_touched_ball_in_friendly_defense_validation.cpp:29]: I expect this function (by the name) to return _true_ if a non-goalie robot touches the ball

(Similar to what Kelvin has mentionsed)

**Review by EvanMorcom — CHANGES_REQUESTED** [review]
This PR is going to bring us some good changes :fire: 🧨 

I've got some comments. Will review again when most the comments are addressed :+1:

**Comment by jared-paul** [comment]
Wait so should I look into using the SSL game-controller or continue with addressing these requested changes?

**Comment by jared-paul** [comment]
@MathewMacDougall @garethellis0  @kelvinkoon @jonathanlew @EvanMorcom

**Comment by MathewMacDougall** [comment]
@jared-paul we're planning to investigate the "CI mode" of the SSL Gamecontroller to see if it's worth integrating. In the meantime we don't want to have you do unneeded work, so there's a new ticket for creating validation functions for the Corner Kick Play that should be pretty similar to what you've done here. We'd like to put this PR on hold until we make a call about the Gamecontroller. (See slack for more details).

**Comment by jonathanlew** [comment]
Closing because we are prioritizing adding SimulatedTests to all plays first

**Pull request closed without merging**
